sqlflow_models/cluster_keras.py

```diff
@@ -0,0 +1,272 @@
+#!usr/bin/env python
+# -*- coding:utf-8 _*-
+
+"""
+__author__ : chenxiang
+__email__ : alfredchenxiang@didichuxing.com
+__file_name__ : cluster_model.py
+__create_time__ : 2019/09/03
+"""
+import tensorflow as tf
+from tensorflow.python import keras
+from tensorflow.python.keras.layers import Dense, Layer
+from tensorflow.python.keras import backend
+import numpy as np
+from sklearn.cluster import KMeans
+from sklearn.metrics import normalized_mutual_info_score, adjusted_rand_score
+from sklearn.utils.linear_assignment_ import linear_assignment
+
+ari = adjusted_rand_score
+nmi = normalized_mutual_info_score
+
+
+def acc(y_true, y_pred):
+    """
+    Calculate clustering accuracy.
+    Using the Hungarian algorithm to solve linear assignment problem.
+    """
+    y_true = y_true.astype(np.int64)
+    assert y_pred.size == y_true.size
+    dims = max(y_pred.max(), y_true.max()) + 1
+    w = np.zeros((dims, dims), dtype=np.int64)
+    for i in range(y_pred.size):
+        w[y_pred[i], y_true[i]] += 1
+
+    ind = linear_assignment(w.max() - w)
+    return sum([w[i, j] for i, j in ind]) * 1.0 / y_pred.size
+
+
+class DECModel(keras.Model):
+    def __init__(self, n_clusters=10,
+                 kmeans_init=20,
+                 run_pretrain=False,
+                 existed_pretrain_model=None,
+                 pretrain_dims=None,
+                 pretrain_activation_func='relu',
+                 pretrain_batch_size=256,
+                 pretrain_epochs=10,
+                 pretrain_initializer='glorot_uniform',
+                 loss=None):
+        """
+        Implement cluster model mostly based on DEC.
```

Comment:
Add a link to DEC?

Reply:
I am not sure that the cluster model is DEC but they are really close. 
I am entangled with whether this link will mislead users.